cpp/tests/strings/floats_tests.cpp

```diff
@@ -128,7 +128,8 @@ TEST_F(StringsConvertTest, ToFloats64)
                                      "456e",
                                      "-1.78e+5",
                                      "-122.33644782",
-                                     "12e+309"};
+                                     "12e+309",
+                                     "1.7976931348623159E308"};
```

Comment:
I worked on this a bit last year but was not able to get consistent results for the following test values compared to `std::stod`
```
"1.7976931348623158E308"
"1.79769313486231581E308"
"-1.7976931348623158E308"
"-1.79769313486231581E308"
```
I would be curious what results you get for these and if you could add them to the gtest.

Reply:
Humm, it seems that my fix still doesn't work right for your examples. I'm looking at the code again.

Reply:
It's weird. I ran a tests with C++ `std::stod`. "1.7976931348623158E308" is converted to `1.7976931348623157081452742e+308`. The maximum allowed double number is also `1.7976931348623157081452742e+308` (the same), which is larger than the number given in the string.

So, somehow, `std::stod` also incorrectly converts string to double.

Reply:
Alright, my fix still produces discrepant results for your examples. The root cause of that is the line:
```
static_cast<FloatType>(digits) * exp10(static_cast<FloatType>(exp_ten))
```
Multiplying two very large floating point numbers leads to the problem. Take the first 2 strings as examples:
 * For the first string ("1.7976931348623158E308"), parsing it we will have `digits = 17976931348623158` and `exp_ten = 292`. Multiplying those number will produce `inf`, which is our result. `std::stod` produces `1.7976931348623157081452742e+308`.

 * For the second string ("1.79769313486231581E308"), parsing it we will have `digits = 179769313486231581` and `exp_ten = 291`. Multiplying those number will produce `1.7976931348623157081452742e+308
` --- equals to the value of `std::numeric_limits<double>::max()`, which is our result. `std::stod` produces `inf`.

I have no idea how to overcome those.


I have no solution for this yet.

Reply:
Yes. I kind of gave up on this for awhile since I'm not sure if it completely satisfies #5225 
I think your solution is much better than the current implementation though.
Could check with the Spark team whether or not this solution works for them?

Reply:
I am now seeing some issues with float parsing. The cudf java unit tests are failing when it tries to parse "3.4028236E38" to a float.  CUDF now returns "3.4028234663852886E38" as the value, but before it would return `Inf` and when I check both java and atof (casting the result to a float from a double) I also get `Inf`.

Reply:
Just be aware that there are also some bugs in the java test for parsing doubles.  Looks like when the test was written someone copied and pasted in the wrong value for max and min double values verifying that the code is wrong. So I will put up a PR to fix that test as soon as this is working too.

Reply:
Can you pull my latest commit and test again please?

Reply:
I upmerged and rebuild everything and I am still getting failures.  The C++ tests fail for me with 

```
[ RUN      ] StringsConvertTest.ToFloats32
../tests/utilities/column_utilities.cu:234: Failure
Failed
differences:
lhs[17] = 3.40282326e+38, rhs[17] = inf

Google Test trace:
../tests/strings/floats_tests.cpp:67:  <--  line of failure

[  FAILED  ] StringsConvertTest.ToFloats32 (0 ms)
```

and the java tests also fail with `"3.4028236E38"` when parsed to a float is not `inf` but is `3.4028232635611926E38` and `"1.7976931348623158E308"` when parsed to a double is not `inf` but is `1.7976931348623157E308`.

Reply:
I tried even more decimal places and did not find a discernible pattern from `std:stod`. 
I don't know you will ever get agreement on floating point values here across C++, CUDA, Java, and Python libraries.

Reply:
I just ran my tests (`float_tests.cpp`) and all tests pass! I'm `make clean` and `make -j` to check again...

Reply:
Fresh build and test. All test passed! "3.4028236E38" is definitely on the list of string=>float32 test.

Boby, can you check if you have pulled my latest commit? Thanks.

Reply:
I am still seeing errors. I just cleaned and rebuilt everything so I am not sure why it is different.

Reply:
I should point out that CI is showing similar failures on some platforms.

Reply:
It seem that the old conversion way (always convert string to double then convert to float) works better as double-precision can handle edge float32 values (near inf) better. I change back my code to use that way, instead of converting string directly to float32. Now all CI tests pass. Boby, I believe that your tests would pass too.

Reply:
Note that parsing string to floats involves some multiplication of the float values. Thus, it's extremely difficult to handle values near inf as for those operations overflow and round-off can happen anytime and lead to invalid results.

Reply:
The latest changes fixed all of my issues

Reply:
Sorry I updated the tests to check other values, and I forgot to update them back. I am still getting issues for double parsing of `"1.7976931348623158E308"`  This is still better then what was before. But it is not 100% perfect.

Reply:
I should say that it is good enough that I am fine with checking it in. It is a corner case that we can document, but it would be nice to fix it.

Reply:
Thanks Boby!!!

I don't have problem with "1.7976931348623158E308" on my computer---forever excuse 😄.

Reply:
Ok, so how is this now different than the original code? 
You are hardcoding `stof<double>` so the template parameter is no longer required.
If this is working better than the change is likely minor.

Reply:
Yes, the change is insignificant. I just tri`ed to improve the original function `atof` by rewriting it a little bit, avoiding round-off on the integer mantissa before converting it to a double number. I also tried to reduce the round-off by moving the exponent around. Again, like @davidwendt did, I found it extremely difficult to correctly handle edge values. 

I also dig into the code of `std::atof` from various `libc` sources---they are huge, and are much more complex, with many macros to handle different platforms and edge cases. So, I  think that I/we may never be able to accurately port `std::atof` to CUDA. Our implementation is good enough.

Reply:
> I also dig into the code of std::atof from various libc sources

Be very careful with looking at any libc implementations. Most of them are GPL and you cannot use anything you find there. It'd be best to avoid looking at them all together to avoid the possibility or appearance of impropriety.

Reply:
> Although my template function is now only used for instantiating double, I prefer to keep it as a template function.

I disagree with this approach. I would rather improve on the magic numbers usage.

Reply:
Ah sure, thanks for let me know that.

Reply:
Ah right. OK, I've changed that---the template function `stof` is now again `stod`.